tabpy-server/tabpy_server/handlers/base_handler.py

```diff
@@ -1,30 +1,143 @@
+        self.request_context_logged = False
+
+    def set_request(self, request: tornado.httputil.HTTPServerRequest):
+        '''
+        Set HTTP(S) request for logger. Headers will be used to
+        append request data as client information, Tableau user name, etc.
+        '''
+        self.remote_ip = request.remote_ip
+        self.method = request.method
+        self.url = request.full_url()
+
+        if 'TabPy-Client' in request.headers:
+            self.client = request.headers['TabPy-Client']
+        else:
+            self.client = None
+
+        if 'TabPy-User' in request.headers:
+            self.tableau_username = request.headers['TabPy-User']
+        else:
+            self.tableau_username = None
+
+    def set_tabpy_username(self, tabpy_username: str):
+        self.tabpy_username = tabpy_username
+
+    def enable_context_logging(self, enable: bool):
+        '''
+        Enable/disable request context information logging.
+
+        Parameters
+        ----------
+        enable: bool
+            If True request context information will be logged and
+            every log entry for a request handler will have call ID
+            with it.
+        '''
+        self.log_request_context = enable
+
+    def _log_context_info(self):
+        if not self.log_request_context:
+            return
+
+        context = f'Call ID: {self.call_id}'
+
+        if self.remote_ip is not None:
+            context += f', Caller: {self.remote_ip}'
+
+        if self.method is not None:
+            context += f', Method: {self.method}'
+
+        if self.url is not None:
+            context += f', URL: {self.url}'
+
+        if self.client is not None:
+            context += f', Client: {self.client}'
 
+        if self.tableau_username is not None:
+            context += f', Tableau user: {self.tableau_username}'
+
+        if self.tabpy_username is not None:
+            context += f', TabPy user: {self.tabpy_username}'
+
+        logging.getLogger(__name__).log(logging.INFO, context)
+        self.request_context_logged = True
+
+    def log(self, level: int, msg: str):
+        '''
+        Log message with or without call ID. If call context is logged and
+        call ID added to any log entry is specified by if context logging
+        is enabled (see CallContext.enable_context_logging for more details).
+
+        Parameters
+        ----------
+        level: int
+            Log level: logging.CRITICAL, ERROR, WARNING, INFO, DEBUG, NOTSET.
+
+        msg: str
+            Message format string.
+
+        args
+            Same as args in Logger.debug().
+
+        kwargs
+            Same as kwargs in Logger.debug().
+        '''
+        extended_msg = msg
+        if self.log_request_context:
+            if not self.request_context_logged:
+                self._log_context_info()
+
+            extended_msg += f', <<call ID: {self.call_id}>>'
+
+        logging.getLogger(__name__).log(level, extended_msg)
```

Comment:
Again would suggest assigning `logger = logging.getLogger(__name__)` in the global namespace; it's done most other places in the code base so would help with consistency

Reply:
This is exactly what I want to avoid - global logger :)